MekHQ/src/mekhq/gui/dialog/HistoricalDailyReportDialog.java

```diff
@@ -0,0 +1,174 @@
+            daysLabel = new JLabel(resourceMap.getString("days.text"));
+            filterPanel = new JPanel();
+            closeBtn = new JButton(resourceMap.getString("closeBtn.text"));
+            cacheInfoLabel = new JLabel(resourceMap.getString("cachedInformationMessage.text"));
+
+            updateLogPanel((Integer)pickTime.getSelectedItem());
+
+            pickTime.addActionListener(new ActionListener() {
+                @Override
+                public void actionPerformed(ActionEvent event) {
+                    updateLogPanel((Integer)pickTime.getSelectedItem());
+                }
+            });
+
+            closeBtn.addActionListener(new ActionListener() {
+                @Override
+                public void actionPerformed(ActionEvent event) {
+                    setVisible(false);
+                }
+            });
+
+            filterPanel.add(pickTimeLabel);
+            filterPanel.add(pickTime);
+            filterPanel.add(daysLabel);
+
+            GridBagConstraints gridBag = new GridBagConstraints();
+            gridBag.fill = GridBagConstraints.HORIZONTAL;
+            gridBag.anchor = GridBagConstraints.NORTHWEST;
+            gridBag.gridx = 0;
+            gridBag.gridy = 0;
+            gridBag.insets = new Insets(15,15,15,15); //add some spacing for readability
+            getContentPane().add(cacheInfoLabel, gridBag);
+
+            gridBag = new GridBagConstraints();
+            gridBag.fill = GridBagConstraints.HORIZONTAL;
+            gridBag.gridx = 0;
+            gridBag.gridy = 1;
+            getContentPane().add(filterPanel, gridBag);
+
+            gridBag = new GridBagConstraints();
+            gridBag.fill = GridBagConstraints.BOTH;
+            gridBag.gridx = 0;
+            gridBag.gridy = 2;
+            gridBag.weightx = 1.0;
+            gridBag.weighty = 1.0;
+            getContentPane().add(logPanel, gridBag);
+
+            gridBag = new GridBagConstraints();
+            gridBag.fill = GridBagConstraints.HORIZONTAL;
+            gridBag.anchor = GridBagConstraints.PAGE_END;
+            gridBag.gridx = 0;
+            gridBag.gridy = 2;
+            getContentPane().add(closeBtn, gridBag);
+        } else {
+            GridBagConstraints gridBag = new GridBagConstraints();
+            gridBag.fill = GridBagConstraints.HORIZONTAL;
+            gridBag.anchor = GridBagConstraints.NORTHWEST;
+            gridBag.gridx = 0;
+            gridBag.gridy = 0;
+            JLabel notice = new JLabel(resourceMap.getString("enableInCampaignOptions.text"));
+            getContentPane().add(notice, gridBag);
+        }
+    }
+
+    private void updateLogPanel(Integer days) {
+        logPanel.clearLogPanel();
+        Date trackDay = null;
+        for (int i = gui.getCampaign().inMemoryLogHistory.size()-1; i >= 0; i--) {
+            long diff = gui.getCampaign().getDate().getTime() - gui.getCampaign().inMemoryLogHistory.get(i).getDate().getTime();
```

Comment:
Avoid indexers like `get(n)` on a linked list, as these must traverse every index less than `n` to get there. Switch to a `for` loop over the elements instead.

Reply:
Hoist `getCampaign().getDate().getTime()` into a local variable outside of the loop.